### PR TITLE
Fix segment functions for models without pre-processing transforms

### LINKED ISSUE #873: Error: "if seg_pair_slice['gt'] is not None and isinstance(seg_pair_slice['gt'][0], list): IndexError: list index out of range"
## Issue description

I tried segmenting an image and received the following error:

```
>>> nii_lst, target_lst = imed_inference.segment_volume(path_model, input_filenames, options=options)
2021-07-23 19:57:12.815 | INFO     | ivadomed.config_manager:_display_differing_keys:150 - Adding the following keys to the configuration file
2021-07-23 19:57:12.815 | INFO     | ivadomed.config_manager:_display_differing_keys:152 - 

2021-07-23 19:57:12.816 | WARNING  | ivadomed.inference:segment_volume:384 - fname_roi has not been specified, then the entire volume is processed.
2021-07-23 19:57:12.954 | INFO     | ivadomed.inference:segment_volume:423 - Loaded 6 axial patches of shape [512, 512].
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/GRAMES.POLYMTL.CA/u097381/miniconda/envs/ads_ivadomed/lib/python3.8/site-packages/ivadomed/inference.py", line 443, in segment_volume
    for i_batch, batch in enumerate(data_loader):
  File "/home/GRAMES.POLYMTL.CA/u097381/miniconda/envs/ads_ivadomed/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 345, in __next__
    data = self._next_data()
  File "/home/GRAMES.POLYMTL.CA/u097381/miniconda/envs/ads_ivadomed/lib/python3.8/site-packages/torch/utils/data/dataloader.py", line 385, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "/home/GRAMES.POLYMTL.CA/u097381/miniconda/envs/ads_ivadomed/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/GRAMES.POLYMTL.CA/u097381/miniconda/envs/ads_ivadomed/lib/python3.8/site-packages/torch/utils/data/_utils/fetch.py", line 44, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/GRAMES.POLYMTL.CA/u097381/miniconda/envs/ads_ivadomed/lib/python3.8/site-packages/ivadomed/loader/mri2d_segmentation_dataset.py", line 181, in __getitem__
    if seg_pair_slice['gt'] is not None and isinstance(seg_pair_slice['gt'][0], list):
IndexError: list index out of range
```

### Current behavior
See error above

### Expected behavior
The code doesn't throw an error

### Steps to reproduce
Model: https://drive.google.com/drive/folders/1vOc-HazqyE4xeyBWwbl4OTpoCcEhafad
Image: image.png in data_test/ in https://osf.io/sv7u2/

```python

from ivadomed import inference as imed_inference

path_model="/home/GRAMES.POLYMTL.CA/u097381/ads_ivadomed/axondeepseg/AxonDeepSeg/models/model_seg_rat_axon-myelin_sem"
input_filenames = ["/home/GRAMES.POLYMTL.CA/u097381/ads_ivadomed/axondeepseg/AxonDeepSeg/models/default_SEM_model/data_test/image.png"]
options = {"pixel_size": 0.13, "overlap_2D":[48,48], "binarize_maxpooling": True}

nii_lst, target_lst = imed_inference.segment_volume(path_model, input_filenames, options=options)
```

Pip freeze:

<details>
```

(ads_ivadomed) u097381@bireli:~/ads_ivadomed/axondeepseg$ pip freeze
absl-py==0.13.0
alabaster==0.7.12
appdirs @ file:///home/conda/feedstock_root/build_artifacts/appdirs_1603108395799/work
argon2-cffi @ file:///home/conda/feedstock_root/build_artifacts/argon2-cffi_1610522574055/work
async-generator==1.10
attrs @ file:///home/conda/feedstock_root/build_artifacts/attrs_1620387926260/work
Babel @ file:///home/conda/feedstock_root/build_artifacts/babel_1619719576210/work
backcall @ file:///home/conda/feedstock_root/build_artifacts/backcall_1592338393461/work
backports.functools-lru-cache @ file:///home/conda/feedstock_root/build_artifacts/backports.functools_lru_cache_1618230623929/work
bids-validator==1.8.0
bleach @ file:///home/conda/feedstock_root/build_artifacts/bleach_1626371669429/work
brotlipy==0.7.0
cachetools==4.2.2
certifi==2021.5.30
cffi @ file:///home/conda/feedstock_root/build_artifacts/cffi_1625835293160/work
chardet @ file:///home/conda/feedstock_root/build_artifacts/chardet_1610093490430/work
charset-normalizer @ file:///home/conda/feedstock_root/build_artifacts/charset-normalizer_1626371162869/work
click==8.0.1
cloudpickle @ file:///home/conda/feedstock_root/build_artifacts/cloudpickle_1598400192773/work
colorama @ file:///home/conda/feedstock_root/build_artifacts/colorama_1602866480661/work
contextlib2 @ file:///home/conda/feedstock_root/build_artifacts/contextlib2_1624848568296/work
coverage @ file:///home/conda/feedstock_root/build_artifacts/coverage_1614551281347/work
cryptography @ file:///home/conda/feedstock_root/build_artifacts/cryptography_1616851476134/work
csv-diff==1.1
cycler==0.10.0
cytoolz==0.11.0
dask @ file:///home/conda/feedstock_root/build_artifacts/dask-core_1625867571991/work
debugpy @ file:///home/conda/feedstock_root/build_artifacts/debugpy_1627074852499/work
decorator @ file:///home/conda/feedstock_root/build_artifacts/decorator_1621187651333/work
defusedxml @ file:///home/conda/feedstock_root/build_artifacts/defusedxml_1615232257335/work
dictdiffer==0.9.0
docopt==0.6.2
docutils @ file:///home/conda/feedstock_root/build_artifacts/docutils_1618676255808/work
entrypoints @ file:///home/conda/feedstock_root/build_artifacts/entrypoints_1605121927639/work/dist/entrypoints-0.3-py2.py3-none-any.whl
et-xmlfile==1.0.1
file-tree @ file:///home/conda/feedstock_root/build_artifacts/file-tree_1626714766742/work
file-tree-fsl @ file:///home/conda/feedstock_root/build_artifacts/file-tree-fsl_1618254561880/work
fsleyes @ file:///home/conda/feedstock_root/build_artifacts/fsleyes_1627037557101/work
fsleyes-props @ file:///home/conda/feedstock_root/build_artifacts/fsleyes-props_1619018550148/work
fsleyes-widgets @ file:///home/conda/feedstock_root/build_artifacts/fsleyes-widgets_1619035512156/work
fslpy @ file:///home/conda/feedstock_root/build_artifacts/fslpy_1624544149005/work
fsspec @ file:///home/conda/feedstock_root/build_artifacts/fsspec_1626188337504/work
future==0.18.2
google-auth==1.33.1
google-auth-oauthlib==0.4.4
grpcio==1.39.0
h5py @ file:///home/conda/feedstock_root/build_artifacts/h5py_1617739432869/work
idna @ file:///home/conda/feedstock_root/build_artifacts/idna_1609836280497/work
imagecodecs @ file:///home/conda/feedstock_root/build_artifacts/imagecodecs_1623996795751/work
imageio @ file:///home/conda/feedstock_root/build_artifacts/imageio_1594044661732/work
imagesize==1.2.0
importlib-metadata @ file:///home/conda/feedstock_root/build_artifacts/importlib-metadata_1625463720459/work
indexed-gzip @ file:///home/conda/feedstock_root/build_artifacts/indexed_gzip_1621853422921/work
iniconfig @ file:///home/conda/feedstock_root/build_artifacts/iniconfig_1603384189793/work
ipykernel @ file:///home/conda/feedstock_root/build_artifacts/ipykernel_1626728085309/work/dist/ipykernel-6.0.3-py3-none-any.whl
ipython @ file:///home/conda/feedstock_root/build_artifacts/ipython_1625027225561/work
ipython-genutils==0.2.0
ipywidgets @ file:///home/conda/feedstock_root/build_artifacts/ipywidgets_1609995587151/work
isodate==0.6.0
ivadomed @ git+git://github.com/ivadomed/ivadomed.git@7a595ae39e94085837582f7176b8ec59bd9a2bc9
jdcal==1.4.1
jedi @ file:///home/conda/feedstock_root/build_artifacts/jedi_1610146791023/work
Jinja2 @ file:///home/conda/feedstock_root/build_artifacts/jinja2_1612119311452/work
joblib @ file:///home/conda/feedstock_root/build_artifacts/joblib_1612898609989/work
jsonschema @ file:///home/conda/feedstock_root/build_artifacts/jsonschema_1614815863336/work
jupyter @ file:///home/conda/feedstock_root/build_artifacts/jupyter_1611871900135/work
jupyter-client @ file:///home/conda/feedstock_root/build_artifacts/jupyter_client_1615693636836/work
jupyter-console @ file:///home/conda/feedstock_root/build_artifacts/jupyter_console_1616560109969/work
jupyter-core @ file:///home/conda/feedstock_root/build_artifacts/jupyter_core_1612125275706/work
jupyterlab-pygments @ file:///home/conda/feedstock_root/build_artifacts/jupyterlab_pygments_1601375948261/work
jupyterlab-widgets @ file:///home/conda/feedstock_root/build_artifacts/jupyterlab_widgets_1609173350931/work
kiwisolver @ file:///home/conda/feedstock_root/build_artifacts/kiwisolver_1610099769230/work
locket==0.2.0
loguru==0.5.3
Markdown==3.3.4
MarkupSafe @ file:///home/conda/feedstock_root/build_artifacts/markupsafe_1621455677251/work
matplotlib @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-suite_1611858708474/work
matplotlib-inline @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-inline_1618935594181/work
mistune @ file:///home/conda/feedstock_root/build_artifacts/mistune_1624941302223/work
more-itertools @ file:///home/conda/feedstock_root/build_artifacts/more-itertools_1621606173336/work
mpld3 @ file:///home/conda/feedstock_root/build_artifacts/mpld3_1624384796275/work
nbclient @ file:///home/conda/feedstock_root/build_artifacts/nbclient_1614336084111/work
nbconvert @ file:///home/conda/feedstock_root/build_artifacts/nbconvert_1624472649437/work
nbformat @ file:///home/conda/feedstock_root/build_artifacts/nbformat_1617383142101/work
nest-asyncio @ file:///home/conda/feedstock_root/build_artifacts/nest-asyncio_1617163391303/work
networkx @ file:///home/conda/feedstock_root/build_artifacts/networkx_1598210780226/work
nibabel @ file:///home/conda/feedstock_root/build_artifacts/nibabel_1606620997676/work
notebook @ file:///home/conda/feedstock_root/build_artifacts/notebook_1621259862661/work
num2words==0.5.10
numpy @ file:///home/conda/feedstock_root/build_artifacts/numpy_1626648380866/work
oauthlib==3.1.1
olefile @ file:///home/conda/feedstock_root/build_artifacts/olefile_1602866521163/work
onnxruntime==1.4.0
opencv-contrib-python==4.5.3.56
opencv-python-headless==4.5.3.56
openpyxl @ file:///home/conda/feedstock_root/build_artifacts/openpyxl_1615404969632/work
packaging @ file:///home/conda/feedstock_root/build_artifacts/packaging_1625323647219/work
pandas==1.3.0
pandocfilters==1.4.2
parse @ file:///home/conda/feedstock_root/build_artifacts/parse_1610612363844/work
parso @ file:///home/conda/feedstock_root/build_artifacts/parso_1617148930513/work
partd @ file:///home/conda/feedstock_root/build_artifacts/partd_1617910651905/work
pathlib2 @ file:///home/conda/feedstock_root/build_artifacts/pathlib2_1625503088473/work
patsy==0.5.1
pexpect @ file:///home/conda/feedstock_root/build_artifacts/pexpect_1602535608087/work
pickleshare @ file:///home/conda/feedstock_root/build_artifacts/pickleshare_1602536217715/work
Pillow @ file:///home/conda/feedstock_root/build_artifacts/pillow_1625677820830/work
pluggy @ file:///home/conda/feedstock_root/build_artifacts/pluggy_1610318077870/work
pooch @ file:///home/conda/feedstock_root/build_artifacts/pooch_1623218528920/work
prettytable @ file:///home/conda/feedstock_root/build_artifacts/prettytable_1614725168556/work
progressbar2 @ file:///home/conda/feedstock_root/build_artifacts/progressbar2_1599661727525/work
prometheus-client @ file:///home/conda/feedstock_root/build_artifacts/prometheus_client_1622586138406/work
prompt-toolkit @ file:///home/conda/feedstock_root/build_artifacts/prompt-toolkit_1623977816122/work
protobuf==3.17.3
ptyprocess @ file:///home/conda/feedstock_root/build_artifacts/ptyprocess_1609419310487/work/dist/ptyprocess-0.7.0-py2.py3-none-any.whl
py @ file:///home/conda/feedstock_root/build_artifacts/py_1607783655754/work
pyasn1==0.4.8
pyasn1-modules==0.2.8
pybids==0.13.1
pycparser @ file:///home/conda/feedstock_root/build_artifacts/pycparser_1593275161868/work
pydicom @ file:///home/conda/feedstock_root/build_artifacts/pydicom_1607562682046/work
Pygments @ file:///home/conda/feedstock_root/build_artifacts/pygments_1620245170812/work
PyOpenGL==3.1.5
pyOpenSSL @ file:///home/conda/feedstock_root/build_artifacts/pyopenssl_1608055815057/work
pyparsing==2.4.7
Pypubsub==4.0.3
PyQt5==5.12.3
PyQt5-sip==4.19.18
PyQtChart==5.12
PyQtWebEngine==5.12.1
pyrsistent @ file:///home/conda/feedstock_root/build_artifacts/pyrsistent_1610146798212/work
PySocks @ file:///home/conda/feedstock_root/build_artifacts/pysocks_1610291447907/work
pytest==6.2.4
pytest-cov @ file:///home/conda/feedstock_root/build_artifacts/pytest-cov_1622651878478/work
python-dateutil @ file:///home/conda/feedstock_root/build_artifacts/python-dateutil_1626286286081/work
python-utils @ file:///home/conda/feedstock_root/build_artifacts/python-utils_1616867218381/work
pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1612179539967/work
PyWavelets @ file:///home/conda/feedstock_root/build_artifacts/pywavelets_1607290812047/work
PyYAML==5.4.1
pyzmq @ file:///home/conda/feedstock_root/build_artifacts/pyzmq_1622038460640/work
qtconsole @ file:///home/conda/feedstock_root/build_artifacts/qtconsole_1625023841584/work
QtPy==1.9.0
raven==6.10.0
requests @ file:///home/conda/feedstock_root/build_artifacts/requests_1626393743643/work
requests-oauthlib==1.3.0
rsa==4.7.2
Rtree @ file:///home/conda/feedstock_root/build_artifacts/rtree_1626987839488/work
scikit-image==0.18.2
scikit-learn @ file:///home/conda/feedstock_root/build_artifacts/scikit-learn_1619628586492/work
scipy @ file:///home/conda/feedstock_root/build_artifacts/scipy_1626684344458/work
seaborn==0.11.1
Send2Trash @ file:///home/conda/feedstock_root/build_artifacts/send2trash_1624366715919/work
six @ file:///home/conda/feedstock_root/build_artifacts/six_1620240208055/work
snowballstemmer @ file:///home/conda/feedstock_root/build_artifacts/snowballstemmer_1611270869511/work
Sphinx @ file:///home/conda/feedstock_root/build_artifacts/sphinx_1626290977112/work
sphinx-rtd-theme @ file:///home/conda/feedstock_root/build_artifacts/sphinx_rtd_theme_1617698677674/work
sphinxcontrib-applehelp==1.0.2
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp @ file:///home/conda/feedstock_root/build_artifacts/sphinxcontrib-htmlhelp_1621704829796/work
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml @ file:///home/conda/feedstock_root/build_artifacts/sphinxcontrib-serializinghtml_1621704833237/work
SQLAlchemy==1.3.24
tabulate @ file:///home/conda/feedstock_root/build_artifacts/tabulate_1614001031686/work
tensorboard==2.5.0
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.0
terminado @ file:///home/conda/feedstock_root/build_artifacts/terminado_1623365314921/work
testpath @ file:///home/conda/feedstock_root/build_artifacts/testpath_1621261527237/work
threadpoolctl @ file:///home/conda/feedstock_root/build_artifacts/threadpoolctl_1626092076920/work
tifffile @ file:///home/conda/feedstock_root/build_artifacts/tifffile_1625322421387/work
toml @ file:///home/conda/feedstock_root/build_artifacts/toml_1604308577558/work
toolz @ file:///home/conda/feedstock_root/build_artifacts/toolz_1600973991856/work
torch==1.5.0
torchvision==0.6.0
tornado @ file:///home/conda/feedstock_root/build_artifacts/tornado_1610094706440/work
tqdm @ file:///home/conda/feedstock_root/build_artifacts/tqdm_1625664838498/work
traitlets @ file:///home/conda/feedstock_root/build_artifacts/traitlets_1602771532708/work
trimesh @ file:///home/conda/feedstock_root/build_artifacts/trimesh_1626884538652/work
urllib3 @ file:///home/conda/feedstock_root/build_artifacts/urllib3_1624634538755/work
wcwidth @ file:///home/conda/feedstock_root/build_artifacts/wcwidth_1600965781394/work
webencodings==0.5.1
Werkzeug==2.0.1
widgetsnbextension @ file:///home/conda/feedstock_root/build_artifacts/widgetsnbextension_1605475534911/work
wxnatpy==0.3.2
wxPython==4.1.1
xarray @ file:///home/conda/feedstock_root/build_artifacts/xarray_1621474818012/work
xnat @ file:///home/conda/feedstock_root/build_artifacts/xnat_1618825969432/work
zipp @ file:///home/conda/feedstock_root/build_artifacts/zipp_1625284368454/work
(ads_ivadomed) u097381@bireli:~/ads_ivadomed/axondeepseg$ 


```
</details>

### Installed packages

<details open>
<summary> Output of <code>pip freeze</code> </summary>

```

```

</details>

### PR DESCRIPTION
**Pull request by mariehbourget**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
Fixes #873, where the command `segment` and the function `segment_volume` would not work with a model without pre-processing because `seg_pair_sclice['gt']` is `[]` instead of `None` in the loader.

I changed the if condition in `getitem` of the classes `MRI2DSegmentationDataset` and `MRI3DSubVolumeSegmentationDataset`, both `seg_pair_sclice['gt'] = []` and `seg_pair_sclice['gt'] = None` are now working with the segment functions.

I added a test for the 2D case i.e. a model without pre-processing, only with `NormalizeInstance`.

I cannot add a test for the 3D case because there are constraints on the size of `LENGTH_3D` and `STRIDE_3D` of the subvolumes [here](https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/loader/mri3d_subvolume_segmentation_dataset.py#L92) which require a `CenterCrop` to a specific size in the pre-processing. So it is not likely that the same issue would happen in 3D but I thought correcting both 2D and 3D was a good idea.

## To test the PR:
- On master branch, train a model without transformations with the following config file (~5 minutes on cpu: this is a test config with the bare minimum, 1 epoch, 1 subject, 1 contrast, etc.) [config_test_873.zip](https://github.com/ivadomed/ivadomed/files/6880921/config_test_873.zip)
- With the same config file, run the `segment` command, you should have the same error as described [here](https://github.com/ivadomed/ivadomed/issues/873#issue-951963063).
- On this branch, run the `segment` command again, the segmentation should run without error.

### CODE REVIEW

**Review by hermancollin — APPROVED** [review]
The timing for this is perfect as I had not yet tried the `segment` command and today I figured I should see if my models generalize well over the whole dataset. I encountered the exact same error as in #873.
With this PR everything is working as expected.

File: testing/functional_tests/test_segment_volume.py [hunk 1]
```diff
@@ -133,6 +133,51 @@ def test_segment_volume_2d(download_functional_test_files):
...
 
+def test_segment_volume_2d_no_prepro_transform(download_functional_test_files):
+    model = imed_models.Unet(in_channel=1,
+                             out_channel=1,
+                             depth=2,
+                             drop_rate=DROPOUT,
+                             bn_momentum=BN)
+
+    if not os.path.exists(PATH_MODEL):
+        os.mkdir(PATH_MODEL)
```
Comment (dyt811) [inline testing/functional_tests/test_segment_volume.py:144]: I would recommend to use either [`os.makedirs`](https://docs.python.org/3/library/os.html#os.makedirs) or even better pathlib's [`mkdir`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.mkdir) (with parents=True) such that if the parent path doesn't exist, they will be created automatically.

File: ivadomed/loader/mri2d_segmentation_dataset.py [hunk 2]
```diff
@@ -178,7 +178,7 @@ def __getitem__(self, index):
             seg_pair_slice, roi_pair_slice = copy.deepcopy(self.indexes[index])
 
         # In case multiple raters
-        if seg_pair_slice['gt'] is not None and isinstance(seg_pair_slice['gt'][0], list):
+        if seg_pair_slice['gt'] and isinstance(seg_pair_slice['gt'][0], list):
```
Comment (dyt811) [inline ivadomed/loader/mri2d_segmentation_dataset.py:181]: Wow. Such small line yet has big impact. A "Truthy" check vs "is None" / is empty array check is really the core of this problem. Thanks for spotting this effectively.

Comment (dyt811) [reply ivadomed/loader/mri2d_segmentation_dataset.py:181]: Also, something else I noticed not from your PR but from this class in general is this habitual direct referencing dict keys for retrieval. I think it is an [possible antipattern](https://deepsource.io/blog/8-new-python-antipatterns/#5-not-using-get-to-return-default-values-from-a-dictionary) that we should address in one of the dev meetings.

**Review by dyt811 — APPROVED** [review]
Expertly done spotting the core issue. Bravo, great job ping point precision fix with new unit test! Hooray. 

Good to merge after branch update.

File: testing/functional_tests/test_segment_volume.py (hunk 1 above)
Comment (mariehbourget) [reply testing/functional_tests/test_segment_volume.py:144]: Good to know! This is a bit out of scope for this PR so I suggest we make the correction to this test and all the other tests from this file in a separate PR.

**Pull request merged**